crates/bounded_planet_core/src/networking/serialization.rs

```diff
@@ -5,40 +5,21 @@ use tracing::trace;
 
 use crate::networking::packets::*;
 
-/// Wrap a stream, ready to send packets which can be decoded by a `BoundedPlanetRecvStream`
-#[derive(Debug)]
-pub struct BoundedPlanetSendStream<T: Session> {
-    send: SendStream<T>,
-}
-
-#[derive(Debug)]
-pub enum SendError {
-    /// Sending a packet failed due to a serialisation error
-    EncodeError(rmp_serde::encode::Error),
-
-    /// Sending a packet failed while writing to the socket
-    WriteError(WriteError),
-}
-
-impl<TSession: Session> BoundedPlanetSendStream<TSession> {
-    pub fn new(send: SendStream<TSession>) -> BoundedPlanetSendStream<TSession> {
-        BoundedPlanetSendStream { send }
-    }
-
-    /// Send a packet over the network
-    pub async fn send_packet(&mut self, packet: &Packet) -> Result<(), SendError> {
+impl Packet {
+    /// Send this packet to a network stream. Should be read at the other end with `Packet::receive(stream)`
+    pub async fn send<T: Session>(&self, stream: &mut SendStream<T>) -> Result<(), SendError> {
         // Encode packet into messagepack format
-        let bytes = rmp_serde::to_vec(&packet).map_err(SendError::EncodeError)?;
+        let bytes = rmp_serde::to_vec(self).map_err(SendError::EncodeError)?;
 
         // Prefix with length (4 bytes, network order)
         let len_bytes = (bytes.len() as u32).to_be_bytes();
```

Comment:
Does messagepack not already provide a mechanism to determine packet length? o_O